### ISSUE TITLE
Reko should respect __thiscall and __stdcall in user-supplied function signatures

### ISSUE THREAD

**Issue by uxmal** [uxmal/reko#320]
Right now, Reko goes off the rails in the subject program `VCSample.exe` because, although an indirect method call has been properly decorated with the right type information, it doesn't respect that the indirect call is of type `__thiscall` and therefore has effects on the stack pointer that standard `__cdecl` calls do not.

@ptomin has proposed a fix that patches the symptoms of this, but it is kind of a hack, and unfortunately is applied to the `TrashedRegisterFinder` class, which will be going away once the `analysis-development`  branch is ready for merging into master. I'd rather fix the core issue, which is that `__thiscall` is not respected. To do this the `IndirectCallRewriter` will need to check to see if the function signature's stackdelta is != 0 and if so, inject an assignment:

``` C
esp = esp + {signature.Stackdelta}
```

Naturally, the X86ProcedureSerializer needs to ensure that `StackDelta` is correctly set when `__thiscall` and `__stdcall` procedures are deserialized.

**Comment by ptomin** [comment]
> @ptomin has proposed a fix that patches the symptoms of this, but it is kind of a hack

Yes, it's a temporary hack and it's not supposed to used when `analysis-development` branch will be merged to `master`. Maybe `analysis-development` will be ready soon and best way is wait for it.

I do not know how new `ssa` works with this case. If it does not it should be fixed off course. Maybe you'll push you last changes to `analysis-development` and I'll see how it works with `VCSample.exe`

As for `master` branch it calls `IndirectCallRewriter` after `TrashedRegisterFinder` so doing stack modification in `IndirectCallRewriter` could not help us to fix it

**Comment by ptomin** [comment]
> Naturally, the X86ProcedureSerializer needs to ensure that StackDelta is correctly set when __thiscall and __stdcall procedures are deserialized.

Is it not already done? Look https://github.com/uxmal/reko/blob/6c5b1fa05943000c2215192262cca7b1bec9a044/src/Arch/X86/X86ProcedureSerializer.cs. I think it should work correctly

**Comment by ptomin** [comment]
> Is it not already done? Look https://github.com/uxmal/reko/blob/6c5b1fa05943000c2215192262cca7b1bec9a044/src/Arch/X86/X86ProcedureSerializer.cs. I think it should work correctly

Not, it works for `__stdcall` but it does not for `__thiscall` and `stdcall`

**Comment by uxmal** [comment]
Right now, the `analysis-development` branch has > 100 broken unit tests. I can push it, if you're OK with me disabling the Travis-CI build for that branch. I have to get it back to stable condition; it's finicky work. Once completed, however, it should be possible to propagate type information from a function signature / globals much sooner in the analysis phase.

**Comment by ptomin** [comment]
> I can push it, if you're OK with me disabling the Travis-CI build for that branch.

I'm OK with it.

**Comment by ptomin** [comment]
> Naturally, the X86ProcedureSerializer needs to ensure that StackDelta is correctly set when __thiscall and __stdcall procedures are deserialized.

Fixed in d5b1b6e

**Comment by uxmal** [comment]
@ptomin : we can close this now?

**Comment by ptomin** [comment]
I think, no. Version in `analysis-branch` does not work properly with this case. I'm going to try to fix it

**Comment by uxmal** [comment]
Ok . Be advised that over  100 unit tests are still broken. I'm working on reducing that number tonight

**Comment by uxmal** [comment]
Right now, I'm investigating how to improve `TrashedRegisterFinder2` so that it can correctly deal with hell nodes. I believe I have something, and will be doing some experiments during the week to confirm.

**Comment by ptomin** [comment]
> 'm investigating how to improve TrashedRegisterFinder2 so that it can correctly deal with hell nodes

What do you mean? I have almost implemented fix to `IndirectCallRewriter` which works on `VCExeSample` in `analysis-development`. But I have not done any changes to `TrashedRegisterFinder2`

**Comment by ptomin** [comment]
Or maybe you are going to make `TrashedRegisterFinder2` works on recursive functions

**Comment by uxmal** [comment]
Yes, I was unclear. The new SSA framework has to be able to cope with recursive function, including clusters of mutually recursive functions like what you'd find in a hand-coded recursive descent parser of any expression language that uses parentheses (look no further than Reko's C parser, which does exactly that).  Right now, it's looking as if I'll repurpose the old TrashedRegisterFinder so that rather than running it on the whole program, it will run on each SCC cluster of mutually dependent functions. Experiments continue...

**Comment by uxmal** [comment]
This has been cleared up in the `analysis-development` branch.
